src/smpp_socket.erl

```diff
@@ -308,17 +303,113 @@ bound(timeout, keepalive, #{role := esme} = State) ->
 bound(timeout, _, State) ->
     reconnect(timeout, ?FUNCTION_NAME, State);
 bound({call, From}, {send_req, Body, Time}, State) ->
-    State1 = send_req(State, Body, From, Time),
-    next_state(?FUNCTION_NAME, State1);
+    case current_time() > Time of
+        true ->
+            State1 = reply(State, {error, sending_expired}, current_time(), From),
+            next_state(?FUNCTION_NAME, State1);
+        false ->
+            case check_limits(State) of
+                {ok, State1} ->
+                    State2 = send_req(State1, Body, From, Time),
+                    next_state(?FUNCTION_NAME, State2);
+                {error, {NextState, State1, Actions}} ->
+                    next_state(NextState, State1, Actions)
+            end
+    end;
 bound(cast, {send_req, Body, Fun, Time}, State) ->
-    State1 = send_req(State, Body, Fun, Time),
-    next_state(?FUNCTION_NAME, State1);
+    case current_time() > Time of
+        true ->
+            State1 = reply(State, {error, sending_expired}, current_time(), Fun),
+            next_state(?FUNCTION_NAME, State1);
+        false ->
+            case check_limits(State) of
+                {ok, State1} ->
+                    State2 = send_req(State1, Body, Fun, Time),
+                    next_state(?FUNCTION_NAME, State2);
+                {error, {NextState, State1, Actions}} ->
+                    next_state(NextState, State1, Actions)
+            end
+    end;
 bound(cast, stop, State) ->
     {stop, normal, State};
 bound(EventType, Msg, State) ->
     State1 = callback(handle_event, EventType, Msg, ?FUNCTION_NAME, State),
     next_state(?FUNCTION_NAME, State1).
 
+-spec rate_limit_cooldown(gen_statem:event_type(), term(), state()) ->
+          gen_statem:event_handler_result(statename()).
+rate_limit_cooldown(info, {tcp, Sock, Data}, #{socket := Sock} = State) ->
+    decode(Data, ?FUNCTION_NAME, State);
+rate_limit_cooldown(info, {tcp_closed, Sock}, #{socket := Sock} = State) ->
+    reconnect(closed, ?FUNCTION_NAME, State);
+rate_limit_cooldown(info, {tcp_error, Sock, Reason}, #{socket := Sock} = State) ->
+    reconnect({inet, Reason}, ?FUNCTION_NAME, State);
+rate_limit_cooldown(timeout, keepalive, #{role := esme} = State) ->
+    State1 = send_req(State, #enquire_link{}),
+    next_state(?FUNCTION_NAME, State1);
+rate_limit_cooldown(timeout, _, State) ->
+    reconnect(timeout, ?FUNCTION_NAME, State);
+rate_limit_cooldown(state_timeout, _, State) ->
+    next_state(bound, State);
+rate_limit_cooldown({call, From}, {send_req, _Body, Time}, State) ->
+    case current_time() > Time of
+        true ->
+            State1 = reply(State, {error, sending_expired}, current_time(), From),
```

Comment:
As we discussed orally: in case a request expires when the socket in _cooldown_ state we should specify the cooldown reason as an error reason in the reply. I.e. the error reason on request expiration in different states:
- `bound`: `sending_expired`
- `rate_limit_cooldown`: `rate_limit`
- `in_flight_limit_cooldown`: `in_flight_limit`